### PR TITLE
 MBS-9632: Convert Stats to React (X) 

### PR DESCRIPTION
**Pull request by reosarevok**
Countries

To be reviewed and merged after https://github.com/metabrainz/musicbrainz-server/pull/964

@yvanzo mentioned having issues here where country.entity was null and it crashed on the "key={country.entity.gid}" of line 69 (https://gist.github.com/yvanzo/78a7a66cd6c181e8ae7034b7aeef0cf6#file-statistics-countries). I haven't seen a case where editor is null, yet dataPoints.length > 0, but since that suggests it can happen, we might want to turn CountryStatT.entity into AreaT | null. I'm still confused whether that should actually ever happen or whether it's a bug somewhere though.

### CODE REVIEW

**Comment by reosarevok** [comment]
I found one issue while testing this: at least for French, table sorting is broken because of the numbers being formatted the French way. The TT file did add_commas() instead of formatCount(), possibly because of this - I'm not familiar enough with tablesorter, do we know if there's a better option than just making the display inconsistent for better sorting?

**Comment by reosarevok** [comment]
Other than that issue (and possibly some eslinting) this should be the end of MBS-9632 (only the timeline is left, which @mwiencek wanted to work on)

**Comment by reosarevok** [comment]
Ignore the tablesorter complaint - got it to work, by adding a space and a dot (French and German separators respectively) to the tablesorter parser. If we have some other language where the number separator is not a space, a dot or a comma, then 1) argh and 2) we can add that too.

File: root/statistics/Countries.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+              ) : formatCount($c, country.label_count)}
+            </td>
+            <td className="t">
+              {formatCount(
+                $c,
+                country.artist_count +
+                country.release_count +
+                country.label_count,
+              )
+              }
```
Comment (yvanzo) [inline root/statistics/Countries.js]: Nit-picking: It should probably on the previous line.

File: root/statistics/Countries.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+          </th>
+          <th>
+            {l('Total')}
+            <div className="arrow" />
+          </th>
+        </tr>
+      </thead>
+      <tbody>
+        {countryStats.map((country, index) => (
+          <tr className={loopParity(index)} key={country.entity.gid}>
```
[the remark this answers is not in the document]
Comment (reosarevok) [reply root/statistics/Countries.js]: Made the same change as we did with editor.

File: root/static/scripts/statistics.js [hunk 3]
```diff
@@ -29,15 +29,22 @@ tablesorter.addWidget({
 
 tablesorter.addParser({
   format: function (s) {
-    return tablesorter.formatFloat(s.replace(/,/g, ''));
+    return tablesorter.formatFloat(s.replace(/(,|\.|\s)/g, ''));
   },
   id: 'fancyNumber',
   is: function (s) {
-    return /^[0-9]?[0-9,.]*$/.test(s);
+    return /^[0-9]?\s?[0-9,.]*$/.test(s);
```
[the remark this answers is not in the document]
Comment (reosarevok) [reply root/static/scripts/statistics.js]: ... actually it seems unneeded. Removed.

**Comment by reosarevok** [comment]
Added flow stuff as well - because flow does not support numbers as keys, I had to turn 1: into [1]: etc though.

File: root/statistics/Countries.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (yvanzo) [reply root/statistics/Countries.js]: Now got `TypeError: Cannot read property 'country_code' of null`

[the remark this answers is not in the document]
Comment (reosarevok) [reply root/statistics/Countries.js]: Check now? :)

**Comment by reosarevok** [comment]
I changed the queries so that country.null stats (which are supposed to exist) actually get collected. The query for releases is fine as-is, I think - ideally the other two queries would either work with containment (so that all artists with an area under the UK are counted for the UK rather than null) or be somehow separated so that "Unknown Country" and "Non-Country Area" are different rows.

File: root/statistics/Countries.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (yvanzo) [reply root/statistics/Countries.js]: Now works :)

**Comment by mwiencek** [comment]
1da64a3 resolves [MBS-9845](https://tickets.metabrainz.org/browse/MBS-9845), right? If so we can add that to the commit message and update the ticket. The commit itself looks good for inclusion, we can make a new ticket for further improvements.

**Comment by mwiencek** [comment]
I created [MBS-10102](https://tickets.metabrainz.org/browse/MBS-10102) for using the area containments for the stats (and assigned to myself since I had something that almost works).

**Pull request merged**